File: pulsar/testing/Tester.py
# For output
from pulsar.output import *
from pulsar.util import line

class Tester:
    def __init__(self, desc):
        self.nfailed = 0
        self.ntest = 1
        self.desc = desc


    def print_header(self):
        print_global_output(line("="))
        print_global_output("{}\n".format(self.desc))
        print_global_output(line("="))



    def print_results(self):
        print_global_output("\n\n")
        if self.nfailed > 0:
            print_global_output(line("="))
            print_global_error("RESULTS: {} failed\n".format(self.nfailed))
            print_global_output(line("="))
        else:
            print_global_output(line("="))
            print_global_success("RESULTS: {} failed\n".format(self.nfailed))
            print_global_output(line("="))
        print_global_output("\n\n")


    def test(self, desc, expected, func, *args):
        """
        Syntax: Test(desc, expected, func, ...)
        desc: The description of the test
        expected: True if the function should pass, False if it should fail
        func(...): The function to call, all values following the function
                   name will be forwarded to the function
        """

        try:
            success = func(*args)
        except Exception as e:
            print_global_debug("Exception: {}\n".format(str(e)))
            success = False
        
        fmt = "{:5} : {:<5}  {:<5}  {:<9}   : {}\n"
        
        passed=(success==expected)
        
        
        print_global_output(fmt.format(self.ntest , expected , success , "Success" if passed else "FAILED", desc))
        if not passed:
            self.nfailed+=1
        
        self.ntest += 1

    def test_value(self, desc, v1, v2, Tol=0.00001):
        if type(v1)==float or type(v2)==float:
            self.test(desc,True,lambda w1, w2: abs(w1-w2)<Tol,v1,v2)
        else:
           self.test(desc, True, lambda w1, w2: w1 == w2, v1, v2)
    
         

File: pulsar/testing/Tester.cpp
/*! \file
 *
 * \brief Definition of the Tester class
 * \author Ryan Richard (ryanmrichard1 <at> gmail.com)
 */

#include "pulsar/testing/Tester.hpp"
#include "pulsar/output/GlobalOutput.hpp"
#include "pulsar/util/StringUtil.hpp"
#include "pulsar/exception/Exceptions.hpp"
#include <cmath>

using pulsar::output::print_global_output;
using pulsar::exception::GeneralException;
using std::to_string;

namespace pulsar {
namespace testing {
//Some utility functions
inline void print_line(){print_global_output(util::line('='));}
inline void print_buffer(){print_global_output("\n\n");}


Tester::Tester(const std::string& desc)
{
    print_line();
    print_global_output(desc+"\n");
    print_line();
}


void Tester::print_results()const
{
    print_buffer();
    print_line();
    const std::string msg="RESULTS: "+to_string(nfailed_)+" failed\n";
    if(nfailed_>0)output::print_global_error(msg);
    else output::print_global_success(msg);
    print_line();
    print_buffer();
    if(nfailed_>0)
        throw GeneralException(to_string(nfailed_)+" tests failed");
}

void Tester::test(const std::string& desc, bool passed)
{
    print_global_output("\n"+to_string(ntest_)+" : "+
             (passed?"Success":"FAILED")+" : "+desc+"\n");
    if(!passed)++nfailed_;
    ++ntest_;
}

void Tester::test(const std::string& desc,double v1,double v2,double tol)
{
    test(desc,std::fabs(v1-v2)<tol);
}


}//End namespace testing 
}//End namespace pulsar

